npm install error when trying to run renovate

On the latest release of renovate, at the `Spawning npm install to create /tmp/tmp-string/package-lock.json` step I'm getting this error:

```
module.js:471
           throw err;
           ^
       
       Error: Cannot find module '/home/renovate/node_modules/.bin/npm'
           at Function.Module._resolveFilename (module.js:469:15)
           at Function.Module._load (module.js:417:25)
           at Module.runMain (module.js:604:10)
           at run (bootstrap_node.js:389:7)
           at startup (bootstrap_node.js:149:9)
           at bootstrap_node.js:504:3
```

I believe it is due to this line here which is referencing the npm binary relatively from the directory the renovate command was executed in:

https://github.com/singapore/renovate/blob/f9b12dbb3eb95b06b4914feb692243dda75b1a40/lib/workers/branch/npm.js#L15

--- a/lib/workers/branch/npm.js
+++ b/lib/workers/branch/npm.js
@@ -1,6 +1,7 @@
 const fs = require('fs-extra');
 const cp = require('child_process');
 const path = require('path');
+const getInstalledPath = require('get-installed-path');
 
 module.exports = {
   generateLockFile,
@@ -12,9 +13,49 @@ async function generateLockFile(tmpDir, logger) {
   let result = {};
   try {
     const startTime = process.hrtime();
-    const npmBin = path.resolve('./node_modules/.bin/npm');
-    const npmOptions = [npmBin, 'install', '--ignore-scripts'];
-    result = cp.spawnSync('node', npmOptions, {
+    let npmCmd = 'node';
+    const npmOptions = [];
+    try {
+      // See if renovate is installed locally
+      npmOptions.push(
+        path.join(
+          await getInstalledPath('npm', {
+            local: true,
+          }),
+          'bin/npm-cli.js'
+        )
+      );
+    } catch (localerr) {
+      logger.debug('No locally installed npm found');
+      // Look inside globally installed renovate
+      try {
+        const renovateLocation = await getInstalledPath('renovate');
+        npmOptions.push(
+          path.join(
+            await getInstalledPath('npm', {
+              local: true,
+              cwd: renovateLocation,
+            }),
+            'bin/npm-cli.js'
+          )
+        );
+      } catch (nestederr) {
+        logger.debug('Could not find globally nested npm');
+        // look for global npm
+        try {
+          npmOptions.push(
+            path.join(await getInstalledPath('npm'), 'bin/npm-cli.js')
+          );
+        } catch (globalerr) {
+          logger.warn('Could not find globally installed npm');
+          npmCmd = 'npm';
+        }
+      }
+    }
+    logger.debug(`Using npm: ${npmOptions[0] || npmCmd}`);
+    npmOptions.push('install');
+    npmOptions.push('--ignore-scripts');
+    result = cp.spawnSync(npmCmd, npmOptions, {
       cwd: tmpDir,
       shell: true,
       env: { NODE_ENV: 'dev', PATH: process.env.PATH },

--- a/lib/workers/branch/npm.js
+++ b/lib/workers/branch/npm.js
@@ -13,11 +13,11 @@ async function generateLockFile(tmpDir, logger) {
   let result = {};
   try {
     const startTime = process.hrtime();
-    let npmCmd = 'node';
-    const npmOptions = [];
+    let cmd = 'node';
+    const options = [];
     try {
       // See if renovate is installed locally
-      npmOptions.push(
+      options.push(
         path.join(
           await getInstalledPath('npm', {
             local: true,
@@ -30,7 +30,7 @@ async function generateLockFile(tmpDir, logger) {
       // Look inside globally installed renovate
       try {
         const renovateLocation = await getInstalledPath('renovate');
-        npmOptions.push(
+        options.push(
           path.join(
             await getInstalledPath('npm', {
               local: true,
@@ -43,19 +43,19 @@ async function generateLockFile(tmpDir, logger) {
         logger.debug('Could not find globally nested npm');
         // look for global npm
         try {
-          npmOptions.push(
+          options.push(
             path.join(await getInstalledPath('npm'), 'bin/npm-cli.js')
           );
         } catch (globalerr) {
           logger.warn('Could not find globally installed npm');
-          npmCmd = 'npm';
+          cmd = 'npm';
         }
       }
     }
-    logger.debug(`Using npm: ${npmOptions[0] || npmCmd}`);
-    npmOptions.push('install');
-    npmOptions.push('--ignore-scripts');
-    result = cp.spawnSync(npmCmd, npmOptions, {
+    logger.debug(`Using npm: ${options[0] || cmd}`);
+    options.push('install');
+    options.push('--ignore-scripts');
+    result = cp.spawnSync(cmd, options, {
       cwd: tmpDir,
       shell: true,
       env: { NODE_ENV: 'dev', PATH: process.env.PATH },

--- a/lib/workers/branch/yarn.js
+++ b/lib/workers/branch/yarn.js
@@ -13,11 +13,11 @@ async function generateLockFile(tmpDir, logger) {
   let result = {};
   try {
     const startTime = process.hrtime();
-    let yarnCmd = 'node';
-    const yarnOptions = [];
+    let cmd = 'node';
+    const options = [];
     try {
       // See if renovate is installed locally
-      yarnOptions.push(
+      options.push(
         path.join(
           await getInstalledPath('yarn', {
             local: true,
@@ -30,7 +30,7 @@ async function generateLockFile(tmpDir, logger) {
       // Look inside globally installed renovate
       try {
         const renovateLocation = await getInstalledPath('renovate');
-        yarnOptions.push(
+        options.push(
           path.join(
             await getInstalledPath('yarn', {
               local: true,
@@ -43,19 +43,19 @@ async function generateLockFile(tmpDir, logger) {
         logger.debug('Could not find globally nested yarn');
         // look for global yarn
         try {
-          yarnOptions.push(
+          options.push(
             path.join(await getInstalledPath('yarn'), 'bin/yarn.js')
           );
         } catch (globalerr) {
           logger.warn('Could not find globally installed yarn');
-          yarnCmd = 'npm';
+          cmd = 'yarn';
         }
       }
     }
-    logger.debug(`Using yarn: ${yarnOptions[0] || yarnCmd}`);
-    yarnOptions.push('install');
-    yarnOptions.push('--ignore-scripts');
-    result = cp.spawnSync(yarnCmd, yarnOptions, {
+    logger.debug(`Using yarn: ${options[0] || cmd}`);
+    options.push('install');
+    options.push('--ignore-scripts');
+    result = cp.spawnSync(cmd, options, {
       cwd: tmpDir,
       shell: true,
       env: { NODE_ENV: 'dev', PATH: process.env.PATH },

--- a/lib/workers/branch/yarn.js
+++ b/lib/workers/branch/yarn.js
@@ -1,22 +1,61 @@
 const fs = require('fs-extra');
 const cp = require('child_process');
 const path = require('path');
+const getInstalledPath = require('get-installed-path');
 
 module.exports = {
   generateLockFile,
 };
 
 async function generateLockFile(tmpDir, logger) {
-  logger.debug('Generating new yarn.lock file');
+  logger.debug(`Spawning yarn install to create ${tmpDir}/yarn.lock`);
   let lockFile = null;
   let result = {};
   try {
     const startTime = process.hrtime();
-    logger.debug(`Spawning yarn install to create ${tmpDir}/yarn.lock`);
-    // Use an embedded yarn
-    const yarnBin = path.resolve('./node_modules/.bin/yarn');
-    const yarnOptions = [yarnBin, 'install', '--ignore-scripts'];
-    result = cp.spawnSync('node', yarnOptions, {
+    let yarnCmd = 'node';
+    const yarnOptions = [];
+    try {
+      // See if renovate is installed locally
+      yarnOptions.push(
+        path.join(
+          await getInstalledPath('yarn', {
+            local: true,
+          }),
+          'bin/yarn.js'
+        )
+      );
+    } catch (localerr) {
+      logger.debug('No locally installed yarn found');
+      // Look inside globally installed renovate
+      try {
+        const renovateLocation = await getInstalledPath('renovate');
+        yarnOptions.push(
+          path.join(
+            await getInstalledPath('yarn', {
+              local: true,
+              cwd: renovateLocation,
+            }),
+            'bin/yarn.js'
+          )
+        );
+      } catch (nestederr) {
+        logger.debug('Could not find globally nested yarn');
+        // look for global yarn
+        try {
+          yarnOptions.push(
+            path.join(await getInstalledPath('yarn'), 'bin/yarn.js')
+          );
+        } catch (globalerr) {
+          logger.warn('Could not find globally installed yarn');
+          yarnCmd = 'npm';
+        }
+      }
+    }
+    logger.debug(`Using yarn: ${yarnOptions[0] || yarnCmd}`);
+    yarnOptions.push('install');
+    yarnOptions.push('--ignore-scripts');
+    result = cp.spawnSync(yarnCmd, yarnOptions, {
       cwd: tmpDir,
       shell: true,
       env: { NODE_ENV: 'dev', PATH: process.env.PATH },

--- a/package.json
+++ b/package.json
@@ -46,6 +46,7 @@
     "commander": "2.11.0",
     "conventional-commits-detector": "0.1.1",
     "fs-extra": "4.0.2",
+    "get-installed-path": "3.0.1",
     "gh-got": "6.0.0",
     "github-url-from-git": "1.5.0",
     "gl-got": "7.0.0",

--- a/test/workers/branch/npm.spec.js
+++ b/test/workers/branch/npm.spec.js
@@ -6,6 +6,8 @@ jest.mock('fs-extra');
 jest.mock('child_process');
 jest.mock('get-installed-path');
 
+getInstalledPath.mockImplementation(() => null);
+
 const fs = require('fs-extra');
 const cp = require('child_process');
 

--- a/test/workers/branch/npm.spec.js
+++ b/test/workers/branch/npm.spec.js
@@ -1,8 +1,10 @@
 const npmHelper = require('../../../lib/workers/branch/npm');
 const logger = require('../../_fixtures/logger');
+const getInstalledPath = require('get-installed-path');
 
 jest.mock('fs-extra');
 jest.mock('child_process');
+jest.mock('get-installed-path');
 
 const fs = require('fs-extra');
 const cp = require('child_process');

--- a/test/workers/branch/yarn.spec.js
+++ b/test/workers/branch/yarn.spec.js
@@ -6,6 +6,8 @@ jest.mock('fs-extra');
 jest.mock('child_process');
 jest.mock('get-installed-path');
 
+getInstalledPath.mockImplementation(() => null);
+
 const fs = require('fs-extra');
 const cp = require('child_process');
 

--- a/test/workers/branch/yarn.spec.js
+++ b/test/workers/branch/yarn.spec.js
@@ -1,8 +1,10 @@
 const yarnHelper = require('../../../lib/workers/branch/yarn');
 const logger = require('../../_fixtures/logger');
+const getInstalledPath = require('get-installed-path');
 
 jest.mock('fs-extra');
 jest.mock('child_process');
+jest.mock('get-installed-path');
 
 const fs = require('fs-extra');
 const cp = require('child_process');

